src/hipscat_import/margin_cache/margin_cache_resume_plan.py

```diff
@@ -0,0 +1,176 @@
+            )
+        self.touch_stage_done_file(self.MAPPING_STAGE)
+
+    def is_mapping_done(self) -> bool:
+        """Are there sources left to count?"""
+        return self.done_file_exists(self.MAPPING_STAGE)
+
+    def get_remaining_map_keys(self):
+        """Fetch a tuple for each pixel/partition left to map."""
+        map_keys = set(self.read_done_keys(self.MAPPING_STAGE))
+        return [
+            (f"{hp_pixel.order}_{hp_pixel.pixel}", hp_pixel)
+            for hp_pixel in self.partition_pixels
+            if f"{hp_pixel.order}_{hp_pixel.pixel}" not in map_keys
+        ]
+
+    @classmethod
+    def reducing_key_done(cls, tmp_path, reducing_key: str):
+        """Mark a single reducing task as done
+
+        Args:
+            tmp_path (str): where to write intermediate resume files.
+            reducing_key (str): unique string for each reducing task (e.g. "3_57")
+        """
+        cls.touch_key_done_file(tmp_path, cls.REDUCING_STAGE, reducing_key)
+
+    def get_remaining_reduce_keys(self):
+        """Fetch a tuple for each object catalog pixel to reduce."""
+        reduced_keys = set(self.read_done_keys(self.REDUCING_STAGE))
+        reduce_items = [
+            (f"{hp_pixel.order}_{hp_pixel.pixel}", hp_pixel)
+            for hp_pixel in self.combined_pixels
+            if f"{hp_pixel.order}_{hp_pixel.pixel}" not in reduced_keys
+        ]
+        return reduce_items
+
+    def is_reducing_done(self) -> bool:
+        """Are there partitions left to reduce?"""
+        return self.done_file_exists(self.REDUCING_STAGE)
+
+    def wait_for_reducing(self, futures):
+        """Wait for reducing stage futures to complete."""
+        self.wait_for_futures(futures, self.REDUCING_STAGE)
+        remaining_sources_to_reduce = self.get_remaining_reduce_keys()
+        if len(remaining_sources_to_reduce) > 0:
+            raise RuntimeError(
+                f"{len(remaining_sources_to_reduce)} reducing stages did not complete successfully."
+            )
+        self.touch_stage_done_file(self.REDUCING_STAGE)
+
+
+def _find_partition_margin_pixel_pairs(stats, margin_order):
+    """Creates a DataFrame filled with many-to-many connections between
+    the catalog partition pixels and the margin pixels at `margin_order`.
+    """
+    norders = []
+    part_pix = []
+    margin_pix = []
+
+    for healpixel in stats:
+        order = healpixel.order
+        pix = healpixel.pixel
+
+        d_order = margin_order - order
+
+        margins = pixel_math.get_margin(order, pix, d_order)
+
+        for m_p in margins:
+            norders.append(order)
+            part_pix.append(pix)
+            margin_pix.append(m_p)
+
+    margin_pairs_df = pd.DataFrame(
+        zip(norders, part_pix, margin_pix),
+        columns=["partition_order", "partition_pixel", "margin_pixel"],
+    )
+    return margin_pairs_df
+
+
+def _create_margin_directory(stats, output_path, storage_options):
```

Comment:
Can we apply the same here?

Reply:
Nah - I just removed the method. The reduce method creates directories as needed, so we don't need to pre-create them.